storage/innobase/row/row0import.cc

```diff
@@ -2140,11 +2193,42 @@ row_import_cleanup(
 		     index = UT_LIST_GET_NEXT(indexes, index)) {
 			index->page = FIL_NULL;
 		}
+
+		prebuilt->trx->rollback();
+	}
+	else
+	{
+	  DBUG_EXECUTE_IF("ib_import_before_commit_crash", DBUG_SUICIDE(););
+	  prebuilt->trx->commit();
 	}
 
-	DBUG_EXECUTE_IF("ib_import_before_commit_crash", DBUG_SUICIDE(););
+	if (fts_table && fts_table != prebuilt->table) {
 
-	prebuilt->trx->commit();
+		if (err == DB_SUCCESS) {
+			/* Reload the table in case of hidden fts column */
+			const table_id_t id = prebuilt->table->id;
+			prebuilt->table->release();
+			dict_sys.remove(prebuilt->table);
+
+			prebuilt->table = dict_table_open_on_id(
+					id, true, DICT_TABLE_OP_NORMAL);
+			prebuilt->table->space = fts_table->space;
+			goto free_fts_table;
+		} else {
+free_fts_table:
+			dict_index_t *index = UT_LIST_GET_FIRST(
+						fts_table->indexes);
+			while (index) {
+				dict_index_t *next_index =
+				  UT_LIST_GET_NEXT(indexes, index);
+				dict_index_remove_from_cache(
+						fts_table, index);
+				dict_mem_index_free(index);
+				index = next_index;
+			}
+			dict_mem_table_free(fts_table);
```

Comment:
```
==154785==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55e711fa0a8c in dict_mem_index_free(dict_index_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/dict/dict0mem.cc:1082:2
    #1 0x55e7119a725e in row_import_cleanup(row_prebuilt_t*, dberr_t, dict_table_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0import.cc:2226:5
    #2 0x55e71199ed1e in row_import_for_mysql(dict_table_t*, row_prebuilt_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0import.cc:4932:9
    #3 0x55e7114ff30d in ha_innobase::discard_or_import_tablespace(char) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/handler/ha_innodb.cc:13417:9
…
  Uninitialized value was created by a heap deallocation
    #0 0x55e70f46a3ec in free (/home/buildbot/amd64-debian-11-msan/build/sql/mariadbd+0x8503ec) (BuildId: fd22fd2a58de3e751344727a72a89bbf6db2a776)
    #1 0x55e711fa08a0 in mem_heap_free(mem_block_info_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/include/mem0mem.inl:419:3
    #2 0x55e711fa08a0 in dict_mem_index_free(dict_index_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/dict/dict0mem.cc:1096:2
```
It’s nice to see that in this special case, MSAN catches the same error as ASAN would:
```
==191432==ERROR: AddressSanitizer: heap-use-after-free on address 0x61600ba70f54 at pc 0x555d6ca84589 bp 0x7f448034aba0 sp 0x7f448034ab98
READ of size 4 at 0x61600ba70f54 thread T10
    #0 0x555d6ca84588 in dict_mem_index_free(dict_index_t*) /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/storage/innobase/dict/dict0mem.cc:1082:2
    #1 0x555d6c727f2e in row_import_cleanup(row_prebuilt_t*, dberr_t, dict_table_t*) /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/storage/innobase/row/row0import.cc:2226:5
    #2 0x555d6c72385a in row_import_for_mysql(dict_table_t*, row_prebuilt_t*) /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/storage/innobase/row/row0import.cc:4932:9
    #3 0x555d6c497271 in ha_innobase::discard_or_import_tablespace(char) /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/storage/innobase/handler/ha_innodb.cc:13417:9
    #4 0x555d6b56daaa in mysql_discard_or_import_tablespace(THD*, TABLE_LIST*, bool) /home/buildbot/amd64-ubuntu-2204-clang14-asan/build/sql/sql_table.cc:5553:35
```